server/service/async/async.go

```diff
@@ -0,0 +1,281 @@
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	redigo "github.com/gomodule/redigo/redis"
+	"github.com/jmoiron/sqlx"
+	"github.com/pkg/errors"
+)
+
+const (
+	labelMembershipHostKeyPattern = "label_membership:{*}"
+	labelMembershipHostKey        = "label_membership:{%d}"
+	labelMembershipReportedKey    = "label_membership_reported:{%d}"
+	collectorLockKey              = "locks:async_collector:{%s}"
+)
+
+type Task struct {
+	Datastore fleet.Datastore
+	Pool      fleet.RedisPool
+	// AsyncEnabled indicates if async processing is enabled in the
+	// configuration. Note that Pool can be nil if this is false.
+	AsyncEnabled bool // TODO: should this be read in a different way, more dynamically, if config changes while fleet is running? Or does that require a restart?
+}
+
+// Collect runs the various collectors as distinct background goroutines if
+// async processing is enabled.  Each collector will stop processing when ctx
+// is done.
+func (t *Task) StartCollectors(ctx context.Context, interval time.Duration, jitterPct int, logger kitlog.Logger) {
+	if !t.AsyncEnabled {
+		level.Debug(logger).Log("task", "async disabled, not starting collectors")
+		return
+	}
+	level.Debug(logger).Log("task", "async enabled, starting collectors")
+
+	labelColl := &collector{
+		name:         "collect_labels",
+		pool:         t.Pool,
+		ds:           t.Datastore,
+		execInterval: interval,
+		jitterPct:    jitterPct,
+		lockTimeout:  time.Minute,
+		handler:      collectLabelQueryExecutions,
+		errHandler: func(name string, err error) {
+			level.Error(logger).Log("err", fmt.Sprintf("%s collector", name), "details", err)
+		},
+	}
+	go labelColl.Start(ctx)
+
+	// log stats at regular intervals
+	// TODO: TBD if it stays that way for prod, but will be useful for load testing
+	go func() {
+		tick := time.Tick(time.Minute)
+		for {
+			select {
+			case <-tick:
+				stats := labelColl.ReadStats()
+				level.Debug(logger).Log("stats", fmt.Sprintf("%#v", stats), "name", labelColl.name)
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (t *Task) RecordLabelQueryExecutions(ctx context.Context, host *fleet.Host, results map[uint]*bool, ts time.Time) error {
+	if !t.AsyncEnabled {
+		host.Modified = true
+		host.LabelUpdatedAt = ts
+		return t.Datastore.RecordLabelQueryExecutions(ctx, host, results, ts)
+	}
+
+	keySet := fmt.Sprintf(labelMembershipHostKey, host.ID)
+	keyTs := fmt.Sprintf(labelMembershipReportedKey, host.ID)
+
+	script := redigo.NewScript(2, `
+    redis.call('ZADD', KEYS[1], unpack(ARGV, 2))
+    return redis.call('SET', KEYS[2], ARGV[1])
+  `)
+
+	// convert results to ZADD arguments, store as -1 for delete, +1 for insert
+	args := make(redigo.Args, 0, 3+(len(results)*2))
+	args = args.Add(keySet, keyTs, ts.Unix())
+	for k, v := range results {
+		score := -1
+		if v != nil && *v {
+			score = 1
+		}
+		args = args.Add(score, k)
```

Comment:
Some pointers in this SO answer: https://stackoverflow.com/questions/42911672/is-there-any-limit-on-the-number-of-arguments-that-redis-commands-such-as-zadd-o.

Basically, given that those are label IDs for a single host, I don't think there's too much cause for concern, although we could add a sanity check on `len(results)` and return an error if it's "unexpectedly large" (to be defined)? I think if there's more than what we can send to a single ZADD, it's probably that there's too many (given our expectations of "much less than 1000"). What we definitely should have higher up, though, is a body size limit to our API endpoint so that suspiciously large requests are rejected, which in turn somewhat limits what we need to worry about inside handlers.

Reply:
Gotcha, yeah, probably not worth while worrying.